Changed lines for successfull MSVC 14.0 compiling

Iceman9 opened the pull request:
Edited lines in ../numpy/core/src/multiarray/mapping.c and ../numpy/core/src/private/mem_overlap.c, namely functions that are not legally defined for MSVC 14.0. Need testing if the free (pointer) functions are properly placed in mem_overlap.c

This fixes the errors when trying to compile with MSVC 14.0 for python 3.5

Commit:
BLD: Enabled building with MSVC 14.0

Reallocated free functions in mem_overlap.c

Cosmetics.

Final indent.

Added tests if pointer==NULL

Fixed indent

Fixed position of goto label.

Fixed ISO C90 violation.

Made simpler checks and removed redundant lines.

njsmith commented:
@Iceman9: Looks like your `free` functions are not in the right place -- all the tests are failing with "*** glibc detected *** python: double free or corruption (fasttop): 0x0000000003e58a50 ***" (see the link below next to "The Travis CI build failed")

Iceman9 commented:
I have looked carefully now and changed the following:
```
        if (diophantine_precompute(n, E, Ep, Gamma, Epsilon)) {
			free(Ep);
			free(Gamma);
			free(Epsilon);
            return MEM_OVERLAP_OVERFLOW;
        }
		return diophantine_dfs(n, n-1, E, Ep, Gamma, Epsilon, b, max_work, require_ub_nontrivial, x,, &count);
}
```
to :
```
        if (diophantine_precompute(n, E, Ep, Gamma, Epsilon)) {
			free(Ep);
			free(Gamma);
			free(Epsilon);
            return MEM_OVERLAP_OVERFLOW;
        }
		res = diophantine_dfs(n, n-1, E, Ep, Gamma, Epsilon, b, max_work, require_ub_nontrivial, x, &count);
		free(Ep);
		free(Gamma);
		free(Epsilon);
        return res;
    }
```

The reason behind it is the ```diophantine_dfs``` function is recursive. I didn't go into depth as to what mathematics is involved I only looked at the structure of the function. Since the previous version simply returned the funcion call with arguments, I determined there is no easy way to somehow detect in the recursive functions where the step ends and returns to the starting point, but I know it returns a value of type ```mem_overlap_t``` or basically an integer from the enumerated list.
So insted returning directly the function call, I stored the result in the same type variable, followed by freeing the malloc-ed pointers and then returning the stored result. 

Hopefully I read the code right now.

charris commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,25 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
         Py_ssize_t count = 0;
 
         if (diophantine_precompute(n, E, Ep, Gamma, Epsilon)) {
+			free(Ep);
```
Indentation looks off, I suspect hard tabs. Numpy C style requires tabs be 4 spaces.

charris commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,25 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
     else {
-        diophantine_term_t Ep[n];
-        npy_int64 Epsilon[n], Gamma[n];
+
```
Why the blank lines?

charris commented:
@pv You might want to check this.

charris commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,25 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
         }
-        return diophantine_dfs(n, n-1, E, Ep, Gamma, Epsilon, b, max_work,
-                               require_ub_nontrivial, x, &count);
+		res = diophantine_dfs(n, n-1, E, Ep, Gamma, Epsilon, b, max_work, require_ub_nontrivial, x, &count);
```
Keep lines < 80 characters.

juliantaylor commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,24 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
+
+        diophantine_term_t *Ep = malloc(n * sizeof(diophantine_term_t));
+        npy_int64 *Epsilon = malloc(n * sizeof(npy_int64));
+        npy_int64 *Gamma = malloc(n *sizeof(npy_int64));
```
could use npy_alloc_cache from alloc.h to avoid some overhead
the value also looks like its bounded by NPY_MAXDIMS, though not using might make removing it in future a tiny bit easier

pv commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,24 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
-        diophantine_term_t Ep[n];
-        npy_int64 Epsilon[n], Gamma[n];
+
+        diophantine_term_t *Ep = malloc(n * sizeof(diophantine_term_t));
```
Needs to check if malloc returned null on error

Iceman9 replied:
Question: Is using assert as in ```assert(Ep!=NULL)`` good enough and acceptable?

pv replied:
asserts may get removed by non-debug builds, so you can as well write it as
```
diophantine_term_t *Ep = NULL;
npy_int64 *Epsilon = NULL, *Gamma = NULL;

res = MEM_OVERLAP_ERROR;
Ep = malloc(...);
if (Ep == NULL) {
    goto done;
}
if (diophantine_precompute(n, E, Ep, Gamma, Epsilon)) {
    res = MEM_OVERLAP_OVERFLOW;
    goto done;
}
...
...
done:
    free(Ep);
    free(Epsilon);
    free(Gamma);
    return res;
```

Iceman9 replied:
Thank you a lot!

Iceman9 commented:
Commited again, please review!

charris commented:
Could you squash the commits together and take a look at `doc/source/dev/gitwash/development_workflow.rst` for writing commit messages? You can use `git rebase -i` for that. If you need help with that, post a comment.

juliantaylor commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,40 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
+        if (Ep == NULL || Epsilon == NULL || Gamma == NULL) {
+            goto done;
+        }
+        if (Epsilon == NULL){
```
this check is redundant

juliantaylor commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,40 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
         }
-        return diophantine_dfs(n, n-1, E, Ep, Gamma, Epsilon, b, max_work,
-                               require_ub_nontrivial, x, &count);
+        if (Gamma == NULL){
```
redundant and empty

juliantaylor commented on numpy/core/src/private/mem_overlap.c:
```diff
@@ -535,15 +536,40 @@ solve_diophantine(unsigned int n, diophantine_term_t *E, npy_int64 b,
+        free(Epsilon);
+        return res;
+
+        done:
```
we do not indent labels and the following code has normal indentation

Iceman9 commented:
@charris Hey, I know this is not the place to go of-topic with this question, but after I squash it into one commit with the standard messaging provided in the doc and made hopefully the last acceptable commit, there was an update in the numpy:numpy master. So I updated my fork:master and then tried the same with my working branch. Because I changed the history with rebase of course I have conflicts, which I cannot resolve. No matter what I do [if I fetch, merge etc] status of my working branch is "updated" but it is not, while my fork:master is updated [even though on my profile:repository isn't]. Basically if I check the log I can see that the latest commits are not there. I did a tmp save before any rebasing or updateing, so the reasonable thing would be to revert to that and then FIRST fetch to update to newest and THEN start to rebase the history [because I did the opposite]?

charris commented:
This is the right place for those questions ;) First, after you squash you will need to do a force push because of the changed history `git push -f origin <whatever>`, and second, if the merge button isn't green because of conflicts, do a rebase on your updated master branch, `git rebase master`. If there are conflicts during the rebase, I believe there is documentation on how to do that as it is a bit tricky, but the short version is, fix the file(s), `git add <the files>`, `git rebase --continue`.

Iceman9 commented:
@pv 
@charris Thanks for you help.
@juliantaylor Noted on the no indentation for labels. I have removed it tough.
I have squashed the commits and made it simple as I should've done it from begining.

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1293,7 +1293,7 @@ _get_field_view(PyArrayObject *arr, PyObject *ind, PyArrayObject **view)
                                     PyArray_SHAPE(arr),
                                     PyArray_STRIDES(arr),
-                                    PyArray_DATA(arr) + offset,
+                                    ((char *)PyArray_DATA(arr)) + offset,
```
`PyArray_Bytes` is a bti nicer here, since that is exactly what it is good for (instead of the cast).

astrofrog commented:
Just wanted to confirm that this works for me and fixes the issue I reported in #6701 - thanks! :+1:

charris commented:
Let's get this in, we can get the `PyArray_BYTES` in another PR.

Thanks @Iceman9 .
